fix(server): fix the scheduler and the scheduler selection logic

Tsukilc opened the pull request:
## Summary

Synced with [apache/hugegraph#3082](https://github.com/apache/hugegraph/pull/3082).

Implements the `ServerInfo` soft-deprecation strategy — retiring HugeGraph's legacy master-worker scheduling mechanism.

## Background

HugeGraph's old `ServerInfoManager` ran a master-worker heartbeat + role-election system on top of the graph backend, independent of PD. This was necessary before PD existed, but today:

- **hstore backend** — PD owns cluster membership, roles, and scheduling. The old server_info was redundant and conflicted with PD.
- **Non-hstore backends** — single-node. There's no master or worker to elect. The mechanism was dead code with backend side-effects.

A concrete symptom: community users reported that docker-compose starts 3 servers but only 1 stays alive — caused by the legacy master-worker heartbeat conflicting with PD in the hstore compose setup.

The solution: **soft-deprecate** `ServerInfoManager` persistence. Stop writing, reading, or deleting server_info in the backend. Keep the in-memory API surface for compatibility. Don't touch historical data.

## Changes

### ServerInfoManager soft-deprecation

| # | Method | Before | After |
|---|--------|--------|-------|
| C1 | `init()` | Created `server_info` schema in backend | No-op (logs deprecation notice) |
| C2 | `initServerInfo()` | Queried DB for conflicts, checked alive master, wrote to backend | Only sets in-memory `globalNodeInfo` |
| C3 | `changeServerRole()` | Updated role in memory + saved to backend | In-memory only |
| C4 | `heartbeat()` | Read/updated/saved timestamp every N seconds | No-op |
| C5 | `close()` | Called `removeSelfServerInfo()` + `systemTransaction()` | Pure in-memory close |

Removed ~322 lines of dead code (`save`, `serverInfo`, `removeServerInfo`, `pickWorkerNode`, `onlySingleNode` field, private helpers). File went from ~451 lines to 129.

### Scheduler simplification

- **`schedulerType()`** — returns `"local"` for non-hstore, `"distributed"` for hstore. The `task.scheduler_type` config key is soft-deprecated: if a user explicitly sets it in their properties file, startup fails fast with a clear error message. Auto-detection replaces manual configuration.
- **`StandardTaskScheduler`** — stripped of multi-node paths (`checkOnMasterNode` reduced to a no-op, `SCHEDULING`/`SCHEDULED` pipeline removed, `pickWorkerNode` removed). Tasks submit immediately as `QUEUED`.
- **`DistributedTaskScheduler`** — cron-based scheduling kept for hstore. Two race conditions fixed:
  - Cancel retry on stale status (re-read from DB)
  - Remove-before-cancel in `cronSchedule()` to prevent concurrent cancel races

### Configuration & startup

- `server.role_election` — soft-deprecated. Setting it to `true` logs a warning and proceeds; role state machine init is skipped.
- Removed the hardcoded `server-1` default for `server.id` — now auto-generated at runtime.

### Review-driven fixes

| Issue | Fix |
|-------|-----|
| Cancel status race — stale prestatus drops the request | Re-read from DB, retry transition to CANCELLING |
| SCHEDULING/SCHEDULED tasks orphaned after upgrade | Include both in `restoreTasks()` |
| `role_election=true` crashes startup | Warn instead of crash |
| `ServerInfoManager.close()` touches backend on shutdown | Pure no-op |
| Test expected `IllegalArgumentException` from `role_election=true` | Updated `testGraphManagerWarnsOnRoleElection` to expect successful construction |
| `testDistributedDeleteKeepsTaskResultRecoverable` — expected DELETING but was SUCCESS | Write DELETING status via direct vertex construction before `deleteFromDB()` so failed result deletions are recoverable |
| `testTask` — expected DELETING but was NEW | Removed DELETING guard from `TaskAndResultScheduler.save()` — it blocked legitimate writes from `updateStatus()`. The resurrect-race is already handled by `HugeTask.cancel()` |
| `testTaskWithoutResult` — expected result but was null | `tasks(ImmutableList.of(id))` defaults to `withResult=false`. Changed to explicit `tasks(ids, true)` |
| Potential NPE in `cronSchedule()` | Replace `containsKey()` + `get()` with `get()` + null-check on `runningTasks` ConcurrentHashMap |
| Redundant DB query in `delete()` | Reuse `task` variable instead of re-querying via `taskWithoutResult()` |
| Thread leak in `TaskSchedulerServerInfoTest` | Wrap `DistributedTaskScheduler` constructor in try block so executors are shut down on failure |
| Commented-out code in `close()` | Removed `//return true;`, fixed `todo` → `TODO` |

### Tests

- New `TaskSchedulerServerInfoTest` — covers `DistributedTaskScheduler` without `ServerInfoManager`, `GraphManager` with `role_election` enabled (warns, doesn't throw), PD peer injection for RocksDB vs hstore.
- `testGraphManagerRejectsRoleElection` renamed to `testGraphManagerWarnsOnRoleElection` — expects construction to succeed instead of throwing.
- `TaskCoreTest` updated for immediate-submit scheduling.

---

Supersedes #2937.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * 调度器类型改为按后端自动选择，旧的调度器类型配置将被忽略并给出弃用提示。
  * 图配置可在未显式配置时继承服务器侧 PD 连接信息。
  * `server.id` 变为可选 legacy id。
* **Bug Fixes**
  * 任务恢复/取消/删除的状态流转更一致，减少中间态与残留异常。
  * 查询无结果任务时改为返回缺失错误（NotFound），而非空结果。
  * 关闭/删除过程更稳健，避免不确定的等待与竞态。
* **Tests**
  * 加强任务与调度器相关清理及断言，提升测试稳定性并覆盖更多场景。
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Tsukilc commented on hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/task/StandardTaskScheduler.java:
```diff
@@ -211,30 +201,9 @@ public <V> Future<?> schedule(HugeTask<V> task) {
+        task.status(TaskStatus.QUEUED);
+        this.save(task);
+        return this.submitTask(task);
     }
```
完成时跳过 decreaseLoad

Tsukilc committed:
fix(server): fix some issues of the distributed scheduler

Tsukilc committed:
fix(server): delete config

zhongbaojun committed:
fix: remove serverid

Tsukilc committed:
fix: fail fast when role-election

Tsukilc committed:
fix: no-ops in ServerInfoManager.Init

Tsukilc committed:
fix: guard pd peers injection

Tsukilc committed:
fix: no-ops in ServerInfoManager.Init

bitflicker64 committed:
fix: address review comments - fail-fast guard, delete race, and test coverage

bitflicker64 committed:
test: update testGraphManagerWarnsOnRoleElection to match new soft-deprecation behavior

bitflicker64 committed:
fix(server): persist DELETING status before deleteFromDB so failed deletes are recoverable

When delete(force=true) is called and deleteFromDB() throws (e.g. result
vertex removal fails), the task must be left as DELETING so cronSchedule()
can retry. Previously the force-delete path skipped the status transition
entirely, leaving the task as SUCCESS — unrecoverable.

bitflicker64 commented:
/gemini help

bitflicker64 commented:
/gemini review

bitflicker64 committed:
fix(server): remove DELETING guard from save(), fix cronSchedule NPE and review issues

- Remove DELETING guard from TaskAndResultScheduler.save() — it blocked
  legitimate DELETING writes from updateStatus(), causing testTask:96
  failure. The resurrect-race is already handled by HugeTask.cancel().
- Write DELETING status before deleteFromDB() in force-delete path so
  failed deletions are recoverable via cronSchedule(). Reuse existing
  task variable instead of re-querying DB.
- Fix potential NPE in cronSchedule(): replace containsKey()+get() with
  get()+null-check on runningTasks ConcurrentHashMap.
- Remove commented-out code and fix todo→TODO in close().
- Fix thread leak in TaskSchedulerServerInfoTest: wrap constructor in
  try block so executors are shut down on failure.

bitflicker64 committed:
fix(test): pass withResult=true to tasks(List) in testTaskWithoutResult

tasks(ImmutableList.of(id)) defaults to withResult=false and returns
tasks without result, but the test expected the result field to be
present. Pass explicit withResult=true to use queryTask(ids) path.

bitflicker64 committed:
fix(test): pass withResult=true to tasks(status, limit, page) too

bitflicker64 commented:
/gemini review

bitflicker64 committed:
fix: add null check for role, sync cancel status overwrite, and Gemini review fixes

- GraphManager.initNodeRole(): add NullPointerException guard on role
  before toUpperCase()
- StandardTaskScheduler.cancel(): overwrite task status to CANCELLED
  on successful cancel for consistency with DistributedTaskScheduler

imbajin commented on hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/StandardHugeGraph.java:
```diff
@@ -226,11 +225,20 @@ public StandardHugeGraph(HugeConfig config) {
         this.name = config.get(CoreOptions.STORE);
+
+        // Fail fast if user still has the removed config key in their properties file
+        if (config.containsKey("task.scheduler_type")) {
```
‼️ Critical Issues

This fail-fast check is too aggressive for phase 1. `task.scheduler_type` was shipped in previous graph config templates, so existing deployments that upgrade in place can still have this key in their `hugegraph.properties`. Even if the legacy scheduler selection must no longer take effect, startup should not fail just because an old no-op key is present.

Please keep this key backward-compatible for now: accept it, log a deprecation warning, ignore its value, and continue with the new backend-based selection (`hstore` -> distributed, others -> local). A hard failure or full removal is better suited for the later disable/delete phases.

imbajin commented on hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/task/DistributedTaskScheduler.java:
```diff
@@ -179,14 +186,20 @@ public void cronSchedule() {
-
+                // Remove BEFORE cancel() so a concurrent cronSchedule()
+                // on another node won't see this task and retry it.
                 runningTasks.remove(cancellingId);
```
‼️ Critical Issues

`runningTasks` is cleared before the task runner has actually stopped. `pendingTasks()` is just `runningTasks.size()`, while `HugeTask.cancel(true)` can return before the callable exits, and `HugeTask` already notes that Gremlin sleep/work may not be interruptible. This lets close/clear/truncate paths observe zero pending tasks while the task thread can still own graph resources and still save progress/result state.

Please keep the task in `runningTasks` until `TaskRunner.finally` runs. Use DB status or the task lock to prevent rescheduling, but let the runner finalization path be the only place that removes the active in-memory task.

imbajin commented on hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/task/DistributedTaskScheduler.java:
```diff
@@ -320,20 +379,27 @@ protected <V> HugeTask<V> deleteFromDB(Id id) {
+        if (task != null && task.status() != TaskStatus.DELETING) {
+            initTaskParams(task);
+            task.overwriteStatus(TaskStatus.DELETING);
+            this.call(() -> {
```
‼️ Critical Issues

The force-delete path marks a freshly loaded task object as `DELETING`, but it does not coordinate with the actual in-memory task stored in `runningTasks`. If that task is still running, this path can remove the task/result vertices while the original task instance later finishes and saves `SUCCESS`, `CANCELLED`, progress, or result data again. The same lifecycle risk also exists when the cron `DELETING` path cancels and physically deletes while the runner is still active.

Please first check and handle the exact running instance for this id: mark it `DELETING`, cancel it, and only physically delete after the runner has exited or after a later cron pass confirms the lock is gone. Add a regression test where a running distributed task is force-deleted and then allowed to finish, and assert the task/result do not reappear.

imbajin commented on hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/task/StandardTaskScheduler.java:
```diff
@@ -146,19 +143,22 @@ private TaskTransaction tx() {
-                    if (selfServer.equals(task.server())) {
-                        taskList.add(task);
-                    }
+                    taskList.add(task);
```
‼️ Critical Issues

Removing the previous `selfServer.equals(task.server())` ownership filter means every local scheduler restores every pending task, and this now also includes historical `SCHEDULING`/`SCHEDULED` tasks. In shared-backend non-hstore deployments, multiple HugeGraph servers can restart and execute the same persisted task because there is no replacement claim/lock.

Please keep a single-owner restore rule, or atomically claim restore candidates before switching them to `RESTORING`. At minimum, do not globally restore old `SCHEDULING`/`SCHEDULED` tasks without a migration path that guarantees exactly one executor.

imbajin commented on hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/task/DistributedTaskScheduler.java:
```diff
@@ -363,6 +429,21 @@ public boolean close() {
+            Future<?> barrier = this.schedulerExecutor.submit(() -> {
+                // pass
+            });
+            barrier.get(schedulePeriod + 5, TimeUnit.SECONDS);
```
⚠️ Important Issues

This barrier only waits behind the scheduler thread, and timeout/interruption only logs a warning before the method closes the task DB transaction. If a cron run just submitted a worker, or a task ignores interrupt, the worker can still run and attempt status/result saves after the task transaction has been closed.

Please stop cron first, drain the scheduler thread, then cancel and wait for `runningTasks` after no new workers can be submitted. If running tasks remain after timeout, close should fail or leave the scheduler registered instead of closing the task DB tx after a warning.

imbajin commented on hugegraph-server/hugegraph-test/src/main/java/org/apache/hugegraph/core/TaskCoreTest.java:
```diff
@@ -76,20 +77,33 @@ public void testTask() throws TimeoutException {
+            HugeTask<?> result = scheduler.delete(id, false);
+            Assert.assertNull(result);
+            HugeTask<?> marked = scheduler.task(id);
+            Assert.assertEquals(TaskStatus.DELETING, marked.status());
```
⚠️ Important Issues

This early return means the distributed scheduler branch only verifies `delete(id, false)` -> `DELETING`, then skips the common task lifecycle assertions below: completion, lookup/listing, completed-task deletion, and `NotFoundException`. Since hstore now auto-selects `DistributedTaskScheduler`, green tests can miss regressions in the main distributed lifecycle.

Please split the incomplete-task delete behavior into a dedicated distributed test, then schedule a fresh task and run the same wait/query/list/delete assertions for both local and distributed schedulers. Also add startup-level coverage that old `task.scheduler_type` configs are accepted as deprecated no-ops during this compatibility phase.

imbajin commented on hugegraph-server/hugegraph-api/src/main/java/org/apache/hugegraph/config/ServerOptions.java:
```diff
@@ -556,9 +556,9 @@ public class ServerOptions extends OptionHolder {
-                    "The id of hugegraph-server.",
-                    disallowEmpty(),
-                    "server-1"
+                    "The id of hugegraph-server, auto-generated if not specified.",
```
⚠️ Important Issues

The option text now says `server.id` is auto-generated when omitted, but the runtime intentionally leaves it empty and the new test asserts that `globalNodeRoleInfo().nodeId()` stays `null`. That mismatch is confusing during this compatibility-sensitive phase because operators may expect a generated id to exist for legacy role/server-info paths.

Please either update the description to say the legacy id is optional/ignored while server-info persistence is disabled, or actually generate and retain an id if any remaining compatibility path still needs one.

imbajin reviewed (commented):
Score: 5.0/10, REQUEST_CHANGES recommended. The direction is right for phase 1: the legacy master-worker mechanism can be neutralized now and removed later. However, this phase still needs to preserve upgrade compatibility and avoid critical runtime regressions: users with old configs must not fail startup after upgrade, and running distributed tasks must not be deleted/cancelled in a way that can resurrect state or race with shutdown.

imbajin committed:
fix(task): bypass legacy scheduler safely

- ignore deprecated task.scheduler_type for upgrade compatibility
- avoid restoring legacy master-worker task statuses
- harden distributed task deletion against save races
- add focused task scheduler regression coverage

imbajin commented:
/gemini review

@coderabbitai full review

imbajin committed:
fix(task): defer deleting running distributed tasks

- avoid physically deleting locally running distributed tasks before runner exit
- keep force delete on running tasks in DELETING and cancel flow
- fail scheduler close when running tasks cannot drain in time
- add hstore regression coverage for force deleting running tasks

imbajin committed:
fix(task): address scheduler review comments

- inject PD peers for hstore backend case-insensitively
- keep queued task persistence after local queue admission
- harden task cleanup and GraphManager test lifecycle
- cover queue-full scheduling and mixed-case hstore config

imbajin commented on hugegraph-server/hugegraph-api/src/main/java/org/apache/hugegraph/core/GraphManager.java:
```diff
@@ -1937,26 +1944,29 @@ public Set<String> getServiceUrls(String graphSpace, String service,
+                    (!"DEFAULT".equals(this.serviceGraphSpace) &&
+                     !graphSpace.equals(this.serviceGraphSpace))) {
+                    return null;
+                }
+                Map<String, Object> config = configs.get(key);
+                String creator = String.valueOf(config.get("creator"));
+                Date createTime = parseDate(config.get("create_time"));
+                Date updateTime = parseDate(config.get("update_time"));
+                HugeGraph graph1 = this.createGraph(graphSpace, name,
+                                                    creator, config, false);
+                graph1.createTime(createTime);
+                graph1.updateTime(updateTime);
+                this.graphs.put(key, graph1);
+                return graph1;
             }
-            Map<String, Object> config = configs.get(key);
-            String creator = String.valueOf(config.get("creator"));
-            Date createTime = parseDate(config.get("create_time"));
-            Date updateTime = parseDate(config.get("update_time"));
-            HugeGraph graph1 = this.createGraph(graphSpace, name,
-                                                creator, config, false);
-            graph1.createTime(createTime);
-            graph1.updateTime(updateTime);
-            this.graphs.put(key, graph1);
-            return graph1;
+            throw new NotFoundException(String.format("Graph '%s' does not exist", name));
```
Leaving this one intentionally out of this patch. The per-key cache-miss race is a valid concurrency concern, but fixing it would expand graph-loading semantics beyond the scheduler cleanup in this PR. I addressed the smaller verified issues in 3a6f20c97 and would track this as a separate follow-up so the current change stays contained.

imbajin commented:
/gemini review

@coderabbitai full review

imbajin committed:
fix(task): prevent task resurrection on delete

- cancel locally running standard tasks before force deletion
- tolerate missing distributed tasks during cancel retry
- add regression coverage for delete and cancel races
